File: app/src/lib/geo/lookup.ts
/**
 * Offline point-in-polygon jurisdiction lookup.
 *
 * Direct port of the proven police-locator algorithm: per-feature bbox
 * prefilter (bboxes are precomputed into the bundle at build time),
 * exact turf.booleanPointInPolygon match, and a nearest-polygon fallback
 * (polygonToLine + pointToLineDistance, 400 m tolerance) for points that
 * miss every polygon because independent boundary simplification opened
 * a hairline gap.
 */
import booleanPointInPolygon from "@turf/boolean-point-in-polygon";
import polygonToLine from "@turf/polygon-to-line";
import pointToLineDistance from "@turf/point-to-line-distance";
import distance from "@turf/distance";
import { point as turfPoint } from "@turf/helpers";
import type { Feature } from "geojson";
import type { FeatureCollection } from "geojson";
import type { Jurisdiction } from "../../types";
import type { GeoPack, GridIndex } from "./geodata";

type Pt = ReturnType<typeof turfPoint>;

/** Grid-index candidate list for a point: the point's cell plus (for the
 *  nearest-fallback's tolerance) the 3×3 neighbourhood. Falls back to
 *  "all features" when the layer ships no grid. */
function gridCandidates(
  features: Feature[],
  grid: GridIndex | undefined,
  [x, y]: number[],
  neighbourhood = 0
): Feature[] {
  if (!grid) return features;
  const c = Math.floor((x - grid.minX) / grid.cell);
  const r = Math.floor((y - grid.minY) / grid.cell);
  const idx = new Set<number>();
  for (let dc = -neighbourhood; dc <= neighbourhood; dc++) {
    for (let dr = -neighbourhood; dr <= neighbourhood; dr++) {
      for (const i of grid.cells[`${c + dc}_${r + dr}`] ?? []) idx.add(i);
    }
  }
  return [...idx].map((i) => features[i]).filter(Boolean);
}

function pointInBbox(
  [x, y]: number[],
  [minX, minY, maxX, maxY]: number[],
  pad = 0
): boolean {
  return (
    x >= minX - pad && x <= maxX + pad && y >= minY - pad && y <= maxY + pad
  );
}

function findContainingFeature(
  features: Feature[],
  pt: Pt
): Feature | null {
  const c = pt.geometry.coordinates;
  for (const f of features) {
    if (f.bbox && !pointInBbox(c, f.bbox)) continue;
    try {
      if (booleanPointInPolygon(pt, f as never)) return f;
    } catch {
      // ignore invalid geometry
    }
  }
  return null;
}

/**
 * Fallback for points that just barely miss every polygon. Picks the
 * polygon whose edge is closest, within `maxKm` of the point.
 */
function nearestPolygon(
  features: Feature[],
  pt: Pt,
  maxKm = 0.4
): Feature | null {
  const c = pt.geometry.coordinates;
  const pad = maxKm / 100; // ~degrees; ~0.004 for 400 m
  let best: Feature | null = null;
  let bestKm = Infinity;
  for (const f of features) {
    if (f.bbox && !pointInBbox(c, f.bbox, pad)) continue;
    try {
      const line = polygonToLine(f as never) as
        | Feature
        | FeatureCollection;
      const lines =
        line.type === "FeatureCollection" ? line.features : [line];
      for (const ln of lines) {
        const d = pointToLineDistance(pt, ln as never, {
          units: "kilometers",
        });
        if (d < bestKm) {
          bestKm = d;
          best = f;
        }
      }
    } catch {
      // ignore
    }
  }
  return bestKm <= maxKm ? best : null;
}

function lookupPolygon(features: Feature[], pt: Pt): Feature | null {
  return findContainingFeature(features, pt) ?? nearestPolygon(features, pt);
}

export interface NearestStation {
  name: string;
  km: number;
}

export interface LookupResult {
  jurisdiction: Jurisdiction;
  /** Matched ward feature — the offline mini-map draws its outline. */
  wardFeature: Feature | null;
  loFeature: Feature | null;
  nearestStation: NearestStation | null;
}

export function lookup(
  pack: GeoPack,
  lat: number,
  lng: number
): LookupResult {
  const pt = turfPoint([lng, lat]);
  const coords = pt.geometry.coordinates;

  // ULBs first (grid-accelerated where the pack ships an index), then
  // village panchayats — statewide TN packs carry every local body, so
  // rural points resolve to their gram panchayat the same way urban ones
  // resolve to their corporation/municipality/town panchayat. A point
  // missing both exactly gets the 400 m nearest-edge fallback across
  // both layers (simplification can open hairline gaps).
  const ulbAll = pack.layers.ulb.features as Feature[];
  const vilAll = (pack.layers.villages?.features ?? []) as Feature[];
  let ulbF = findContainingFeature(
    gridCandidates(ulbAll, pack.grids?.ulb, coords),
    pt
  );
  if (!ulbF && vilAll.length) {
    ulbF =
      findContainingFeature(
        gridCandidates(vilAll, pack.grids?.villages, coords),
        pt
      ) ??
      nearestPolygon(gridCandidates(ulbAll, pack.grids?.ulb, coords, 1), pt) ??
      nearestPolygon(
        gridCandidates(vilAll, pack.grids?.villages, coords, 1),
        pt
      );
  }
  const loF = lookupPolygon(pack.layers.lo.features as Feature[], pt);
  const trF = lookupPolygon(pack.layers.traffic.features as Feature[], pt);

  const jurisdiction: Jurisdiction = { scope: "out" };
  const up = (ulbF?.properties ?? {}) as Record<string, string>;
  const lp = (loF?.properties ?? {}) as Record<string, string | boolean>;

  if (ulbF) {
    jurisdiction.scope = "in";
    jurisdiction.corporation = up.corp;
    jurisdiction.city = up.city;
    jurisdiction.ward = up.ward;
    jurisdiction.wardName = up.wardName;
    jurisdiction.zone = up.zone;
    // village panchayats + cantonments carry these instead of ward/zone
    jurisdiction.block = up.block;
    jurisdiction.district = up.district;
  } else if (lp.avadi === true) {
    // Avadi Corporation has no published ward polygons — the only
    // geometry covering it is the L&O commissionerate layer. Claim it
    // only when no local-body polygon matched, so neighbouring
    // municipalities aren't mislabelled.
    jurisdiction.scope = "in";
    jurisdiction.corporation = "Avadi Corporation";
    jurisdiction.city = "Avadi";
    jurisdiction.wardPending = true;
  }

  // Police jurisdictions render whenever their polygons match — the
  // data is honest about itself (no polygon → no claim).
  if (loF) {
    jurisdiction.scope = "in";
    const p = lp as Record<string, string>;
    jurisdiction.loStation = p.station;
    const zoneShort = (p.zone ?? "").replace(/^chennai\s+/i, "");
    jurisdiction.loMeta = [p.ac, p.dc, zoneShort].filter(Boolean).join(" · ");
    jurisdiction.loPhone = p.phone;
  }
  if (trF) {
    jurisdiction.scope = "in";
    const p = trF.properties as Record<string, string>;
    jurisdiction.trafficStation = p.station;
    jurisdiction.trafficMeta = [p.subDivision, p.district]
      .filter(Boolean)
      .join(" · ");
  }

  // Nearest station point — where the pack ships station points.
  let nearest: NearestStation | null = null;
  if (jurisdiction.scope === "in" && pack.layers.stations.features.length) {
    let bestKm = Infinity;
    let bestName = "";
    for (const f of pack.layers.stations.features) {
      const d = distance(pt, f as never, { units: "kilometers" });
      if (d < bestKm) {
        bestKm = d;
        bestName =
          (f.properties as Record<string, string>).label ??
          (f.properties as Record<string, string>).name ??
          "";
      }
    }
    if (bestName) nearest = { name: bestName, km: bestKm };
  }

  return {
    jurisdiction,
    wardFeature: ulbF,
    loFeature: loF,
    nearestStation: nearest,
  };
}

File: native-android/app/src/main/java/city/reclaimchennai/gpscam/geo/Lookup.kt
package city.reclaimchennai.gpscam.geo

import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.hypot
import kotlin.math.PI

/**
 * Offline point-in-polygon jurisdiction lookup — port of the proven web
 * implementation: per-feature bbox prefilter, even-odd polygon test
 * (holes included), and a nearest-polygon-edge fallback (400 m) for
 * hairline gaps opened by boundary simplification.
 */

data class Jurisdiction(
    val inScope: Boolean = false,
    val corporation: String? = null,
    val city: String? = null,
    val ward: String? = null,
    val wardName: String? = null,
    val zone: String? = null,
    val wardPending: Boolean = false,
    val loStation: String? = null,
    val trafficStation: String? = null,
)

data class LookupResult(
    val jurisdiction: Jurisdiction,
    val wardFeature: GeoFeature?,
    val loFeature: GeoFeature?,
)

object Lookup {

    fun lookup(pack: GeoPack, lat: Double, lng: Double): LookupResult {
        val ulbF = findContaining(pack.ulb.features, lng, lat)
        val loF = findContaining(pack.lo.features, lng, lat)
            ?: nearestPolygon(pack.lo.features, lng, lat)
        val trF = findContaining(pack.traffic.features, lng, lat)
            ?: nearestPolygon(pack.traffic.features, lng, lat)

        var j = Jurisdiction()
        if (ulbF != null) {
            val p = ulbF.props
            j = j.copy(
                inScope = true,
                corporation = p["corp"] as? String,
                city = p["city"] as? String,
                ward = p["ward"] as? String,
                wardName = p["wardName"] as? String,
                zone = p["zone"] as? String,
            )
        } else if (loF?.props?.get("avadi") == true) {
            // Avadi has no published ward polygons — only claim it when no
            // local-body polygon matched (mirror of the web rule)
            j = j.copy(
                inScope = true,
                corporation = "Avadi Corporation",
                city = "Avadi",
                wardPending = true,
            )
        }
        if (loF != null) {
            j = j.copy(inScope = true, loStation = loF.props["station"] as? String)
        }
        if (trF != null) {
            j = j.copy(inScope = true, trafficStation = trF.props["station"] as? String)
        }
        return LookupResult(j, ulbF, loF)
    }

    private fun findContaining(
        features: List<GeoFeature>, x: Double, y: Double
    ): GeoFeature? {
        for (f in features) {
            val b = f.bbox
            if (b != null && (x < b[0] || x > b[2] || y < b[1] || y > b[3])) continue
            if (contains(f, x, y)) return f
        }
        return null
    }

    /** even-odd test across every ring of every polygon of the feature */
    private fun contains(f: GeoFeature, x: Double, y: Double): Boolean {
        for (rings in f.polygons) {
            if (rings.isEmpty()) continue
            if (!ringContains(rings[0], x, y)) continue
            var inHole = false
            for (h in 1 until rings.size) {
                if (ringContains(rings[h], x, y)) { inHole = true; break }
            }
            if (!inHole) return true
        }
        return false
    }

    private fun ringContains(ring: DoubleArray, x: Double, y: Double): Boolean {
        var inside = false
        var j = ring.size / 2 - 1
        for (i in 0 until ring.size / 2) {
            val xi = ring[i * 2]; val yi = ring[i * 2 + 1]
            val xj = ring[j * 2]; val yj = ring[j * 2 + 1]
            if ((yi > y) != (yj > y) &&
                x < (xj - xi) * (y - yi) / (yj - yi) + xi
            ) inside = !inside
            j = i
        }
        return inside
    }

    /** closest polygon edge within maxKm — equirectangular approximation */
    private fun nearestPolygon(
        features: List<GeoFeature>, x: Double, y: Double, maxKm: Double = 0.4
    ): GeoFeature? {
        val pad = maxKm / 100.0
        var best: GeoFeature? = null
        var bestKm = Double.MAX_VALUE
        val kx = 111.32 * cos(y * PI / 180.0) // km per deg lng
        val ky = 110.57                        // km per deg lat
        for (f in features) {
            val b = f.bbox
            if (b != null &&
                (x < b[0] - pad || x > b[2] + pad || y < b[1] - pad || y > b[3] + pad)
            ) continue
            for (rings in f.polygons) for (ring in rings) {
                var j2 = ring.size / 2 - 1
                for (i in 0 until ring.size / 2) {
                    val d = segDistKm(
                        x, y,
                        ring[j2 * 2], ring[j2 * 2 + 1],
                        ring[i * 2], ring[i * 2 + 1],
                        kx, ky
                    )
                    if (d < bestKm) { bestKm = d; best = f }
                    j2 = i
                }
            }
        }
        return if (bestKm <= maxKm) best else null
    }

    private fun segDistKm(
        px: Double, py: Double,
        ax: Double, ay: Double, bx: Double, by: Double,
        kx: Double, ky: Double,
    ): Double {
        // project into local km space, then classic point-segment distance
        val pX = px * kx; val pY = py * ky
        val aX = ax * kx; val aY = ay * ky
        val bX = bx * kx; val bY = by * ky
        val dx = bX - aX; val dy = bY - aY
        val len2 = dx * dx + dy * dy
        val t = if (len2 == 0.0) 0.0
        else (((pX - aX) * dx + (pY - aY) * dy) / len2).coerceIn(0.0, 1.0)
        return hypot(pX - (aX + t * dx), pY - (aY + t * dy))
    }
}

/** Zone/ward display formatting — same conventions as the web app:
 *  number first, name in brackets. */
object GeoFormat {
    private val GCC_ZONES = mapOf(
        "thiruvottriyur" to 1, "manali" to 2, "madhavaram" to 3,
        "tondiarpet" to 4, "royapuram" to 5, "thiru-vika-nagar" to 6,
        "ambattur" to 7, "anna nagar" to 8, "teynampet" to 9,
        "kodambakkam" to 10, "valasarvakkam" to 11, "alandur" to 12,
        "adyar" to 13, "perungudi" to 14, "shozhanganallur" to 15,
    )

    fun ward(w: String?): String {
        if (w.isNullOrBlank()) return ""
        val n = w.trimStart('0')
        return n.ifEmpty { w }
    }

    fun zone(zRaw: String?): String {
        if (zRaw.isNullOrBlank()) return ""
        if (zRaw.startsWith("borough", ignoreCase = true)) return zRaw
        var raw = zRaw.replace(Regex("\\s+zone$", RegexOption.IGNORE_CASE), "").trim()
        raw = raw.replace(Regex("^zone\\s*", RegexOption.IGNORE_CASE), "").trim()
        Regex("^(\\d+)(?:\\s+(.+))?$").find(raw)?.let { m ->
            val num = m.groupValues[1].toInt()
            val name = m.groupValues[2].trim()
            return if (name.isEmpty()) "Zone $num" else "Zone $num ($name)"
        }
        Regex("^(.+?)\\s*\\((\\d+)\\)$").find(raw)?.let { m ->
            return "Zone ${m.groupValues[2].toInt()} (${m.groupValues[1].trim()})"
        }
        GCC_ZONES[raw.lowercase()]?.let { return "Zone $it ($raw)" }
        return "Zone $raw"
    }
}
